extensions/aws/processors/PutS3Object.cpp

```diff
@@ -0,0 +1,487 @@
+    ->withDescription("Specifies the algorithm used for server side encryption.")
+    ->build());
+const core::Property PutS3Object::ProxyHost(
+  core::PropertyBuilder::createProperty("Proxy Host")
+    ->withDescription("Proxy host name or IP")
+    ->supportsExpressionLanguage(true)
+    ->build());
+const core::Property PutS3Object::ProxyPort(
+  core::PropertyBuilder::createProperty("Proxy Port")
+    ->withDescription("The port number of the proxy host")
+    ->supportsExpressionLanguage(true)
+    ->build());
+const core::Property PutS3Object::ProxyUsername(
+    core::PropertyBuilder::createProperty("Proxy Username")
+    ->withDescription("Username to set when authenticating against proxy")
+    ->supportsExpressionLanguage(true)
+    ->build());
+const core::Property PutS3Object::ProxyPassword(
+  core::PropertyBuilder::createProperty("Proxy Password")
+    ->withDescription("Password to set when authenticating against proxy")
+    ->supportsExpressionLanguage(true)
+    ->build());
+
+const core::Relationship PutS3Object::Success("success", "FlowFiles are routed to success relationship");
+const core::Relationship PutS3Object::Failure("failure", "FlowFiles are routed to failure relationship");
+
+void PutS3Object::initialize() {
+  // Set the supported properties
+  std::set<core::Property> properties;
+  properties.insert(ObjectKey);
+  properties.insert(Bucket);
+  properties.insert(ContentType);
+  properties.insert(AccessKey);
+  properties.insert(SecretKey);
+  properties.insert(CredentialsFile);
+  properties.insert(AWSCredentialsProviderService);
+  properties.insert(StorageClass);
+  properties.insert(Region);
+  properties.insert(CommunicationsTimeout);
+  properties.insert(FullControlUserList);
+  properties.insert(ReadPermissionUserList);
+  properties.insert(ReadACLUserList);
+  properties.insert(WriteACLUserList);
+  properties.insert(CannedACL);
+  properties.insert(EndpointOverrideURL);
+  properties.insert(ServerSideEncryption);
+  properties.insert(ProxyHost);
+  properties.insert(ProxyPort);
+  properties.insert(ProxyUsername);
+  properties.insert(ProxyPassword);
+  setSupportedProperties(properties);
+  // Set the supported relationships
+  std::set<core::Relationship> relationships;
+  relationships.insert(Failure);
+  relationships.insert(Success);
+  setSupportedRelationships(relationships);
+}
+
+minifi::utils::optional<Aws::Auth::AWSCredentials> PutS3Object::getAWSCredentialsFromControllerService(const std::shared_ptr<core::ProcessContext> &context) const {
+  std::string service_name;
+  if (context->getProperty(AWSCredentialsProviderService.getName(), service_name) && !service_name.empty()) {
+    std::shared_ptr<core::controller::ControllerService> service = context->getControllerService(service_name);
+    if (nullptr != service) {
+      auto aws_credentials_service = std::static_pointer_cast<minifi::aws::controllers::AWSCredentialsService>(service);
```

Comment:
is `static_pointer_cast` a safe choice here? we are fetching (instantiating maybe?) a ControllerService based on a configuration parameter, could that contain a value that results in a ControllerService other than AWSCredentialsService?

Reply:
That can be a possible user error in case the user gives the name of another service here. It can cause an error which could be avoided in this case, I'll use a dynamic cast here instead.

Reply:
Fixed in [2e71a70](https://github.com/apache/nifi-minifi-cpp/pull/915/commits/2e71a707a31f5564360c40c7bfdd38c3d97302ef)